plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -411,6 +432,48 @@ isFirstVarBeforeEffects n t = do
   go (Unknown : _) = False
   go [] = False
 
+{-| Check if all the variable occurrences are "delayed".
+This means that a variable is used in either of the following ways:
+  * inside a 'delay' term
+  * inside a lambda body
+  * inside a case branch
+-}
+isVarDelayed
+  :: forall name uni fun a
+   . (InliningConstraints name uni fun)
+  => name
+  -> Term name uni fun a
+  -> Maybe Bool
+isVarDelayed name term = do
+  case go False [] term of
+    [] -> Nothing
+    xs -> Just (and xs)
+ where
+  go :: Bool -> [Bool] -> Term name uni fun a -> [Bool]
+  go delayed usages = \case
+    Var _ann name' ->
+      if name == name'
+        then delayed : usages
+        else usages
+    LamAbs _ann _paramName body ->
+      go True usages body
+    Apply _ann t1 t2 ->
+      go delayed usages t1 ++ go delayed usages t2
+    Force _ann t ->
+      go delayed usages t
+    Delay _ann t ->
+      go True usages t
```

Comment:
I think this function is more complex than necessary. You don't need to recurse here if you encouter a `Delay`. We don't care about what's inside `Delay`.

You can make this function return `True` if at least one occurrence is _not_ delayed. Then, avoid descending into subterms of Delay, Lambda and Case branches. Then in the Var case, if it is match, just return True.

Reply:
> You don't need to recurse here if you encouter a Delay.
Good point, will do.